src/hooks/useSelector.js

```diff
@@ -12,61 +11,70 @@ function useSelectorWithStoreAndSubscription(
   store,
   contextSub
 ) {
-  const [, forceRender] = useReducer(s => s + 1, 0)
+  const [state, setState] = useState(() => ({
+    storeState: store.getState(),
+    selector,
+    selectedState: selector(store.getState())
+    // subscriptionCallbackError: undefined
+  }))
 
   const subscription = useMemo(() => new Subscription(store, contextSub), [
     store,
     contextSub
   ])
 
-  const latestSubscriptionCallbackError = useRef()
-  const latestSelector = useRef()
-  const latestSelectedState = useRef()
-
-  let selectedState
+  let selectedState = state.selectedState
 
   try {
-    if (
-      selector !== latestSelector.current ||
-      latestSubscriptionCallbackError.current
-    ) {
-      selectedState = selector(store.getState())
-    } else {
-      selectedState = latestSelectedState.current
+    if (selector !== state.selector || state.subscriptionCallbackError) {
+      const newSelectedState = selector(state.storeState)
+      if (!equalityFn(newSelectedState, selectedState)) {
+        selectedState = newSelectedState
+        // schedule another update
+        setState(prevState => ({
+          ...prevState,
+          selector,
+          selectedState
+          // subscriptionCallbackError: undefined
+        }))
+      }
     }
   } catch (err) {
-    if (latestSubscriptionCallbackError.current) {
-      err.message += `\nThe error may be correlated with this previous error:\n${latestSubscriptionCallbackError.current.stack}\n\n`
+    if (state.subscriptionCallbackError) {
+      err.message += `\nThe error may be correlated with this previous error:\n${state.subscriptionCallbackError.stack}\n\n`
     }
 
     throw err
   }
 
-  useIsomorphicLayoutEffect(() => {
-    latestSelector.current = selector
-    latestSelectedState.current = selectedState
-    latestSubscriptionCallbackError.current = undefined
-  })
-
-  useIsomorphicLayoutEffect(() => {
+  useEffect(() => {
     function checkForUpdates() {
-      try {
-        const newSelectedState = latestSelector.current(store.getState())
-
-        if (equalityFn(newSelectedState, latestSelectedState.current)) {
-          return
+      const newStoreState = store.getState()
+      setState(prevState => {
+        let newSelectedState
+        let subscriptionCallbackError
+        try {
+          newSelectedState = prevState.selector(newStoreState)
+
+          if (equalityFn(newSelectedState, prevState.selectedState)) {
```

Comment:
potential staleness issue here: if the store state changes, but the selected state does not, then in line 30 you may be using a stale state if the component gets rendered due to an outside effect

Reply:
Oh, this one I missed. Absolutely.

Only the solution I can think of again is to useCallback. OK, that was difficult.